docs/proposals/component-registration.md

```diff
@@ -0,0 +1,409 @@
+  - Similar to Pods
+    - Listable, like most other top-level API resources
+    - Have both Spec and Status (filterable on both/either)
+    - Spec would have (optional) LivenessProbe and ReadinessProbe for determining the condition
+    - Describes a single instance
+    - Can be selected by labels, metadata, spec, and status
+    - Is namespaced (components would use a specific namespace, likely "kube-system")
+    - Status would have an array of [Conditions](../devel/api-conventions.md#typical-status-properties)
+    - Can be named with a name generator (similar to pods created by replication controllers) when instances of the same type don't already have unique IDs (generated name is returned in the create response)
+  - Similar to Endpoints (current External Endpoints usage)
+    - Spec would have address (ip or host) and ports (optionally named) to connect with an external service provider instance
+    - Manually added for non-component External Service provider instances (e.g. Oracle DB) that do not know about k8s (rather than adding Endpoints)
+  - Different from Endpoints
+    - Singular, instead of plural (use -List, namespacing, and labels for grouping)
+    - Includes both Ready & NotReady endpoints
+    - Supports multiple and future conditions (in an extensible Conditions array)
+2. Modify ServiceSpec.Selector to select from both Pods and ExternalServiceProviders
+3. Add a new ExternalServiceProvider Controller that handles probing ExternalServiceProviders
+  - Probe results update ExternalServiceProvider.Status.Conditions
+4. Modify Endpoints Controller to manage ExternalServiceProvider Endpoints
+  - Readiness from ExternalServiceProvider.Status.Conditions analogous to PodStatus.Conditions
+5. Modify the Primary Components to self-register ExternalServiceProviders (one per component instance) instead of an Endpoint
+  - Group ExternalServiceProviders (component instances) with labels (e.g. "type=component", "component-type=core")
+6. Modify the Addon Components' Pod and Service registration to use similar labels (e.g. "type=component", "component-type=addon")
+7. Register component Services (with ServiceSpec.Selector) as part of cluster deployment
+  - Services of new components can also be added at runtime, before any instances exist. Each instance would self-register a ExternalServiceProviders.
+8. Deprecate `/componentsstatuses` in favor of requesting a filtered list of ExternalServiceProviders
+  - ExternalServiceProviders are external-only, but have generic conditions (liveness and readiness)
+  - Endpoints are external + internal, but only have readiness (for now)
+
+
+#### API Types
+
+The following is a draft of how the API models might change.
+
+```
+type ExternalServiceProvider struct {
+	TypeMeta
+	ObjectMeta
+	Spec ExternalServiceProviderSpec
+	Status ExternalServiceProviderStatus
+}
+
+type ExternalServiceProviderSpec struct {
+	Host string
+	Ports []ExternalServiceProviderPort
+	LivenessProbe *Probe
```

Comment:
I'm ok with `ReachabilityProbe`, but I think there's some nuances worth discussing.

For example, there a fine line between doing a TCP probe locally and a TCP probe remotely. Both test that something is listening on a port, and due to the complexities of containerized networking and potential overlays, both may also test IP routing and port mapping.

For all we know, the ExternalServiceProvider could be running on one of the k8s nodes, just not managed by k8s, or in another VM on the same host. The exact implication of the probe result is always going to depend on the topology. 

There's also a real problem of perspective, because the ExternalServiceProvider might be reachable from the (external or internal) controller doing the checking and still not be reachable by an (internal) Pod.

It's also worth pointing out that the LivenessProbe is deliberately optional. They only need to supply one if they intend to use the result. How the result is used is entirely external to k8s at this point. If the operator didn't have a better more local method, they could use it as a way to decide when to restart the ExternalServiceProvider. This might be useful for external components, for example, for people that don't have a system like borg monitoring them.

As for "per-port readiness probes and conditions", I'm happy to keep this API in sync with whatever happens to Pods. One of the reasons to keep them similar is just to keep cognitive overhead low.